src/Workspaces/Remote/Core/Diagnostics/PerformanceTrackerService.cs

```diff
@@ -0,0 +1,402 @@
+            }
+
+            OnSnapshotAdded();
+        }
+
+        public void GenerateReport(List<BadAnalyzerInfo> badAnalyzers)
+        {
+            using (var pooledRaw = SharedPools.Default<Dictionary<string, (double mean, double stddev)>>().GetPooledObject())
+            {
+                var rawPerformanceData = pooledRaw.Object;
+
+                lock (_gate)
+                {
+                    // first get raw aggregated peformance data from the queue
+                    _queue.GetPerformanceData(rawPerformanceData);
+                }
+
+                // make sure there are some data
+                if (rawPerformanceData.Count == 0)
+                {
+                    return;
+                }
+
+                using (var generator = new ReportGenerator(this, MinLOFValue, MeanThreshold, StddevThreshold, badAnalyzers))
+                {
+                    generator.Report(rawPerformanceData);
+                }
+            }
+        }
+
+        private void RecordBuiltInAnalyzers(IEnumerable<AnalyzerPerformanceInfo> snapshot)
+        {
+            foreach (var entry in snapshot)
+            {
+                _builtInMap[entry.AnalyzerId] = entry.BuiltIn;
+            }
+        }
+
+        private bool AllowTelemetry(string analyzerId)
+        {
+            if (_builtInMap.TryGetValue(analyzerId, out var builtIn))
+            {
+                return builtIn;
+            }
+
+            return false;
+        }
+
+        private void OnSnapshotAdded()
+        {
+            SnapshotAdded?.Invoke(this, EventArgs.Empty);
+        }
+
+        private sealed class ReportGenerator : IDisposable, IComparer<BadAnalyzerInfo>
+        {
+            private readonly double _minLOFValue;
+            private readonly double _meanThreshold;
+            private readonly double _stddevThreshold;
+
+            private readonly PerformanceTrackerService _owner;
+            private readonly List<BadAnalyzerInfo> _badAnalyzers;
+            private readonly PooledObject<List<IDisposable>> _pooledObjects;
+
+            public ReportGenerator(
+                PerformanceTrackerService owner,
+                double minLOFValue,
+                double meanThreshold,
+                double stddevThreshold,
+                List<BadAnalyzerInfo> badAnalyzers)
+            {
+                _pooledObjects = SharedPools.Default<List<IDisposable>>().GetPooledObject();
+
+                _owner = owner;
+
+                _minLOFValue = minLOFValue;
+                _meanThreshold = meanThreshold;
+                _stddevThreshold = stddevThreshold;
+
+                _badAnalyzers = badAnalyzers;
+            }
+
+            public void Report(Dictionary<string, (double mean, double stddev)> rawPerformanceData)
+            {
+                // convert string (analyzerId) to index
+                var analyzerIdIndex = GetAnalyzerIdIndex(rawPerformanceData.Keys);
```

Comment:
@roofkitty can you take a look at this code where I implemented your Lof algorithm? also, can you point me wiki url for references about Lof, K distances and neighbor theories?

Reply:
all other code is just a plumbing to move data to this and report results somewhere.

Reply:
Here is wiki to Lof https://en.wikipedia.org/wiki/Local_outlier_factor. It also has definition for K distance and neighbor.